system/protocols/generic/factory.py

```diff
@@ -0,0 +1,225 @@
+# coding=utf-8
+import importlib
+
+from twisted.internet import reactor
+from twisted.internet.protocol import ClientFactory
+
+from system.protocols.generic.protocol import Protocol
+from system.logging.logger import getLogger
+
+__author__ = 'Gareth Coles'
+
+
+class BaseFactory(ClientFactory):
+    """
+    A Twisted factory, for producing protocols. Now with less fuckery!
+
+    Subclass this in your protocol's *factory.Factory* class, and use it to
+    describe how your protocol is built and how it connects. If you find this
+    confusing, you may want to read the other docstrings in this class, and
+    take a look at the factory for the *irc* protocol.
+    """
+
+    protocol = None
+    protocol_module = None
+
+    reconnection_attempts = 0
+    reconnection_config = None
+
+    shutting_down = False
+
+    def __init__(self, protocol_name, config, factory_manager):
+        self.name = protocol_name
+        self.config = config
+        self.factory_manager = factory_manager
+
+        self.logger = getLogger("F: {}".format(self.name))
+        self.reconnection_config = factory_manager.main_config["reconnections"]
```

Comment:
\<Kasen\> reconnection settings should really be protocol specific (but fall back to global settings)
\<Kasen\> it also needs a `max_delay` setting